### PR TITLE
[Prim][PIR] Sink decomp graph

### PR DESCRIPTION
**Pull request by cyber-pioneer**
### PR types
New features

### PR changes
Others

### Description
Pcard-66975
Sink decomp mechanism into cpp. In future, decomp mechanism will run in cpp by default.

### CODE REVIEW

**Review by cxxly — DISMISSED** [review]
LGTM with TODO:

- 下沉机制成熟之后，sink_decomp -> decomp，移除旧的代码

**Comment by cyber-pioneer** [comment]
> LGTM with TODO:
> 
> * 下沉机制成熟之后，sink_decomp -> decomp，移除旧的代码

ok

File: paddle/fluid/primitive/base/decomp_trans.cc [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+      op_info.GetInterfaceImpl<paddle::dialect::DecompInterface>();
+  if (decomp_interface_impl == nullptr) return false;
+  return true;
+}
+
+bool DecompProgram::check_decomp_dynamic_shape(pir::Operation* op) {
+  for (auto item : op->operands()) {
+    auto value = item.source();
+    // check if initialized in case of optional input.
+    if (value.impl() && value.type().storage()) {
```
Comment (Charles-hit) [inline paddle/fluid/primitive/base/decomp_trans.cc]:
```suggestion
    if (IsEmptyValue(value)) {
```

Comment (cyber-pioneer) [reply paddle/fluid/primitive/base/decomp_trans.cc]: done

**Review by Charles-hit — APPROVED** [review]

**Review by Aurelius84 — APPROVED** [review]
LGTM

**Pull request merged**